dnssec-tools/validator/doc/libval_check_conf.1
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "LIBVAL_CHECK_CONF 1"
.TH LIBVAL_CHECK_CONF 1 "2011-06-30" "perl v5.12.4" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
libval_check_conf \- command\-line program for checking validity of the
validator configuration files.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    libval_check_conf [options]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This program checks a given validator configuration file for any syntax errors. 
If no file is specified as a command line option, the default \fBdnsval.conf\fR 
file is checked.
.SH "RETURN VALUES"
.IX Header "RETURN VALUES"
The program returns 0 on success and \-1 on failure.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-d, \-\-dnsval\-conf" 4
.IX Item "-d, --dnsval-conf"
Specifies the location for the dnsval.conf file.  If this option is not
specified the default dnsval.conf file recognized by the validator library is
used instead.
.IP "\-r, \-\-resolv\-conf" 4
.IX Item "-r, --resolv-conf"
Specifies the location for the resolv.conf file.  If this option is not
specified the default resolv.conf file recognized by the validator library is
used instead.
.IP "\-i, \-\-root\-hints" 4
.IX Item "-i, --root-hints"
Specifies the location for the root.hints file.  If this option is not
specified the default root.hints file recognized by the validator library is
used instead.
.IP "\-v, \-\-verbose" 4
.IX Item "-v, --verbose"
Displays detailed messages and warnings associated with the validator configuration file check process.
.IP "\-h, \-\-help" 4
.IX Item "-h, --help"
Displays a usage help message and exits the program.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2005\-2011 \s-1SPARTA\s0, Inc.  All rights reserved.
See the \s-1COPYING\s0 file included with the DNSSEC-Tools package for details.
.SH "AUTHORS"
.IX Header "AUTHORS"
Suresh Krishnaswamy (hserus@users.sourceforge.net)
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fB\f(BIdnsval.conf\fB\|(3)\fR


dnssec-tools/NEWS
This file contains a list of major changes per release.  See the
ChangeLog file for a complete set of changes and their details
details.


1.10
 - New Features:
    - New Apps:     (see the validator/apps directory for details)
                    - dnssec-check: check dnssec support from your ISP
                    - dnssec-nodes: graphically displays a DNS
		      hierarchy, color coded by each node's DNSSEC status
                    - dnssec-system-tray: displays pop-up
     		      notifications when a libval-enabled application
		      triggers a DNSSEC error
                    - lookup: a graphical DNS lookup utility that
		      displays the results in a hierarchical tree and
		      color codes the window according to DNSSEC status

    - libval:       - Added support for building on Windows.
                    - added support for falling back to recursion when
                      the caching name server does not appear to 
                      support DNSSEC. This also works as a mechanism
                      to work around poisoned or misbehaving cache. 
                    - Significant improvements to the the asynchronous support. 
    - lsdnssec:     - Improvements to lsdnssec to display different
                      output depending on whether a zone is a
                      stand-alone zone or under control of rollerd.
    - nagios:       - Plugins for the nagios monitoring system which
                      enable monitoring of zone rollerover states.
    - firefox:      - Improved patches that work with the most recent firefox

 Plus many more minor features and bug fixes

1.9
 - New Features:
    - lsdnssec:     - Added a new flag (-p) to show only zones in a
                      particular rollerd phase.
                    - fixed bugs to align timing output with rollerd.
    - rollerd:      - Added a -logtz flag for logging timezones
                    - fixed bugs related to the -alwayssign flag.
		    - zonesigner's path is taken from the config file.
    - rollctl:      - Added -rollall and -rollzone options.
    - zonesigner:   - Assumes keys need to be generated for new zones
		      (Assumes -genkeys option was given if a keyrec file
		      can't be found.)
                    - Exits with unique exit codes if a failure occurs.
                      ("zonesigner -xc CODE" can lookup a description for it.")
		    - Added the -phase option so rollover options could be
		      more easily specified.
    - lights:       - A simple GUI to check the status of rollover states
    - blinkenlights:- Added hide/show commands for rollrec names and zone
		      names, for split-zone support
    - cleankrf:	    - Fixed deletion of obsolete set keyrecs.
    - GUI commands: - Fixed how the Exit command works so they don't coredump.

    - libsres
      & libval:     - New beta support for issuing asynchronous requests.
                      This can speed up queries by up to 4 times if used.
                      (see example code in validator/apps/validator_selftest.c)
                    - NSEC3, DLV and IPv6 are enabled by default.
                    - improved logging and logging-callback support.
    - drawvalmap    - Can output PNG files now

 - Packaging:
                    - Our download page now allows you to download
                      the C validator libraries independently of the
                      full DNNSEC-Tools tool-suite.

 - Many bugs were also fixed in the 240+ changes.

1.8
 - New Features:
    - zonesigner, rollerd
                    - Made changes so that these tools are more compatible 
                      with recent versions of Bind
                    - The zone_errors configuration parameter allows a zone-
                      specific maximum to be set.  Once exceeded, that zone
                      will be skipped rather than allowing rollover to continue.
    - blinkenlights
                    - Recognizes when rollerd abruptly quits, so error messages
                      aren't spewed interminably.
    - ZonFile::Fast - Fixed parsing of DS records containing spaces and 
                      parsing of mname and rname SOA fields
                    - Added support for parsing KEY records
    - keyrec.pm     - Made changes to properly lock keyrec files before 
                      writing to them.
                    - Begun process of deprecating keyrec_open().
    - mapper:       - added a new option: --node-size for mapping
                      complex zones.
    - dnspktflow:   - added two new options:
                      --layout-style for selecting the layout style to use
                      --node-size for mapping complex zones.
                    - Add new (default) option to cluster
                      authoritative nodes together to help better
                      understand the relationships between traffic
                      patterns and authoritative name server/zone arrangement.
    - libval:       - Now distributed with the Root TA.
                    - Added stricter checks for openssl SHA-256 support in 
                      configure. 
                    - Added several improvements that allow the validator to 
                      lookup information within provably insecure zones that 
                      do not handle EDNS0 requests nicely. This includes 
                      adding support for turning off EDNS0 when traversing a 
                      name hierarchy that leads to a provably insecure zone, 
                      EDNS0 fallback support, and additional checks to check 
                      the sanity of response data.
                    - Fixed certain bugs in CNAME handling and in the 
                      validation of proofs accompanying wildcard responses, 
                      referrals and alias chains.
                    - Fixed support for RSADSA and RSASHA-512 signature 
                      validation.

dnssec-tools/validator/apps/libval_check_conf.c
/*
 * Copyright 2005-2011 SPARTA, Inc.  All rights reserved.
 * See the COPYING file distributed with this software for details.
 */

/*
 * Program to check if the provided dnsval.conf file is valid 
 *
 */
#include "validator-config.h"
#include <validator/validator.h>

#ifdef HAVE_LIBGEN_H
#include <libgen.h>
#endif

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#ifdef HAVE_GETOPT_LONG
static struct option prog_options[] = {
    {"help", 0, 0, 'h'},
    {"dnsval-conf", 1, 0, 'd'},
    {"resolv-conf", 1, 0, 'r'},
    {"root-hints", 1, 0, 'i'},
    {"verbose", 0, 0, 'v'},
    {0, 0, 0, 0}
};
#endif

void usage(char *progname)
{
    printf("Usage: %s [options] \n", progname);
    printf("Check if the provided dnsval.conf file is syntactically valid.\n");
    printf("Primary Options:\n");
    printf("        -h, --help                        Display this help and exit\n");
    printf("        -d, --dnsval-conf=<dnsval.conf>   Specify the dnsval.conf file\n");
    printf("        -r, --resolv-conf=<resolv.conf>   Specify the resolv.conf file\n");
    printf("        -i, --root_h-nts=<root.hints>     Specify the root.hints file\n");
    printf("        -v, --verbose                     Enable verbose mode\n");
}

int main(int argc, char *argv[])
{
    int             retval;
    int             c;
    const char     *args = "hvd:r:i:";
    char           *progname;
    val_log_t      *logp;
    val_context_t *context = NULL;
    char           *dnsval_conf = NULL;
    char           *resolv_conf = NULL;
    char           *root_hints = NULL;
    int            debug = 0;

    progname = basename(argv[0]);

    while (1) {

#ifdef HAVE_GETOPT_LONG
        int             opt_index = 0;
#ifdef HAVE_GETOPT_LONG_ONLY
        c = getopt_long_only(argc, argv, args, prog_options, &opt_index);
#else
        c = getopt_long(argc, argv, args, prog_options, &opt_index);
#endif
#else                           /* only have getopt */
        c = getopt(argc, argv, args);
#endif
    
        if (c == -1) {
            break;  
        }

        switch (c) {
            case 'h':   
                usage(progname);
                return (0);

            case 'v':
                debug=1;
                logp = val_log_add_optarg("7:stdout", 1);
                if (NULL == logp) { /* err msg already logged */
                    usage(progname);
                    return (-1);
                }
                break;

            case 'd':
                dnsval_conf = optarg;
                break;

            case 'r':
                resolv_conf = optarg;
                break;

            case 'i':
                root_hints = optarg;
                break;

            default: 
                fprintf(stderr, "Unknown option %s (c = %d [%c])\n",
                        argv[optind - 1], c, (char) c);
                usage(progname);
                return (-1);
        }
    }

    if (debug == 0) {
        logp = val_log_add_optarg("5:stdout", 1);
    }

    if (dnsval_conf == NULL && debug) {
        fprintf(stdout, "dnsval.conf is not specified. Using system defined dnsval.conf for libval.\n");
    }
    
    if (resolv_conf == NULL && debug) {
        fprintf(stdout, "resolv.conf is not specified. Using system defined resolv.conf for libval.\n");
    }
    
    if (root_hints == NULL && debug) {
        fprintf(stdout, "root.hints is not specified. Using system defined root-hints for libval.\n");
    }
    retval = val_create_context_with_conf(NULL, dnsval_conf, resolv_conf, root_hints, &context);

    if (retval != VAL_NO_ERROR) {
        fprintf(stdout, "Result: FAILED. %s \n", p_val_err(retval)); 
        return (-1);
    }

    fprintf(stdout, "Result : OK. \n");
    val_free_context(context);
    return 0;
}
